presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java

```diff
@@ -151,40 +170,45 @@ public OperatorContext getOperatorContext()
     @Override
     public void finish()
     {
+        ListenableFuture<?> currentlyBlocked = blocked;
+        statisticAggregation.finish();
+        ListenableFuture<?> blockedOnAggregation = statisticAggregation.isBlocked();
+        ListenableFuture<?> blockedOnFinish = NOT_BLOCKED;
         if (state == State.RUNNING) {
             state = State.FINISHING;
             finishFuture = pageSink.finish();
-            blocked = toListenableFuture(finishFuture);
+            blockedOnFinish = toListenableFuture(finishFuture);
             updateWrittenBytes();
         }
+        if (currentlyBlocked.isDone() && blockedOnAggregation.isDone() && blockedOnFinish.isDone()) {
+            this.blocked = NOT_BLOCKED;
+        }
+        else {
+            this.blocked = allAsList(currentlyBlocked, blockedOnAggregation, blockedOnFinish);
+        }
     }
 
     @Override
     public boolean isFinished()
     {
-        updateBlockedIfNecessary();
-        return state == State.FINISHED && blocked == NOT_BLOCKED;
+        return blocked.isDone() && state == State.FINISHED;
```

Comment:
nit: reverse (as it was originally): `state == State.FINISHED && ..` -- do the cheaper thing first

Reply:
I didn't see any shortpath for the all finished case? Is there one?